wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/wsnext/WorkspaceNextObjectsRetriever.java

```diff
@@ -0,0 +1,299 @@
+        if (serviceReference.getParameters().isEmpty()) {
+          continue;
+        }
+
+        String serviceName = serviceReference.getName();
+        String serviceVersion = serviceReference.getVersion();
+
+        for (CheServiceParameter cheServiceParameter : serviceReference.getParameters()) {
+          // add dollar sign and curly brackets because parameter is easier to find with these signs
+          // in the map keys
+          String parameterKey =
+              serviceName + "/" + serviceVersion + "/${" + cheServiceParameter.getName() + "}";
+          parameters.putIfAbsent(parameterKey, new ArrayList<>());
+          List<String> cheServiceParameters = parameters.get(parameterKey);
+          cheServiceParameters.add(cheServiceParameter.getValue());
+        }
+      }
+    }
+    return parameters;
+  }
+
+  private CheService getService(String serviceName, String serviceVersion)
+      throws InfrastructureException {
+    try {
+      URI getServiceURI = featureApi.clone().path("service").path(serviceName).path(serviceVersion)
+          .build();
+
+      return getBody(getServiceURI, CheService.class);
+    } catch (IllegalArgumentException | UriBuilderException e) {
+      LOG.error(e.getLocalizedMessage(), e);
+      throw new InternalInfrastructureException(
+          format("Service %s/%s retrieval failed", serviceName, serviceVersion));
+    }
+  }
+
+  @VisibleForTesting
+  protected <T> T getBody(URI uri, Class<T> clas) throws InfrastructureException {
+    HttpURLConnection httpURLConnection = null;
+    try {
+      httpURLConnection = (HttpURLConnection) uri.toURL().openConnection();
+
+      int responseCode = httpURLConnection.getResponseCode();
+      if (responseCode != 200) {
+        throw new InfrastructureException(
+            format("Can't get object by URI '%s'. Error: %s", uri.toString(),
+                getError(httpURLConnection)));
+      }
+
+      return parseResponseStreamAndClose(httpURLConnection.getInputStream(), clas);
+    } catch (IOException e) {
+      throw new InfrastructureException(e.getMessage(), e);
+    } finally {
+      if (httpURLConnection != null) {
+        httpURLConnection.disconnect();
+      }
+    }
+  }
+
+  private String getError(HttpURLConnection httpURLConnection) throws IOException {
+    try (InputStreamReader isr = new InputStreamReader(httpURLConnection.getInputStream())) {
+      return CharStreams.toString(isr);
+    }
+  }
+
+  protected <T> T parseResponseStreamAndClose(InputStream inputStream, Class<T> clazz)
+      throws IOException {
+    try (InputStreamReader reader = new InputStreamReader(inputStream)) {
+      T objectFromJson = GSON.fromJson(reader, clazz);
+      if (objectFromJson == null) {
+        LOG.error(
+            "Response doesn't contain any data, though it was expected to contain some.");
+        throw new IOException(
+            "Internal server error. Unexpected response body received.");
```

Comment:
I would say that `empty response body` would be more accurate message.

Reply:
Does message `Response contains an empty body, though it was expected to contain data` looks better to you?

Reply:
If the message will contain that this response is from "Feature API" then it would be even more clear and helpful.